Add verified revocation pair

indomitableSwan opened the pull request:
Fixes #168.

marsella commented on zkabacus-crypto/src/revlock.rs:
```diff
@@ -168,52 +216,78 @@ impl RevocationLockCommitment {
+                lock: rp.lock,
+            })
+            .unwrap();
+            let maybe_rp_from_secret = RevocationPair::try_from(maybe_rs).unwrap();
```
Nit on naming: I would not call these three things `maybe_`s, because you're actively unwrapping them. There is no situation where they are maybe `Some` and maybe `None`: they're either a `RevocationPair` or the program panics.

The behavior is fine, but I would remove the `maybe_`s from the names.

marsella commented on zkabacus-crypto/src/revlock.rs:
```diff
@@ -222,20 +296,91 @@ mod test {
+
+    #[test]
+    // Check deserialization of revocation pairs from pair
+    pub fn checked_deserialization_from_pair_works() {
```
Nit: this test doesn't explicitly test deserialization, it tests the validation functions. Can you change the name and comment? I would expect a deserialization test to have something like 
```
y = bincode::deserialize(bincode::serialize(x));
assert_eq!(y, x)
```

I would suggest writing such a test anyway (and possibly manually messing up the serialized thing) to make sure the validation functions are called correctly.

indomitableSwan replied:
OK, tried to address. My tests are a little repetitive now, but I'm not sure it makes sense to combine them.

marsella reviewed (changes requested):
Mid-dev review: there's one important issue with the privacy of the `RevocationPair` fields. The rest of the comments are relatively small.

indomitableSwan commented on zkabacus-crypto/src/revlock.rs:
```diff
@@ -16,55 +16,48 @@ use {
+#[allow(missing_copy_implementations)]
+pub struct RevocationPair {
+    /// A revocation lock.
+    pub lock: RevocationLock,
+    /// The associated revocation secret.
+    pub secret: RevocationSecret,
```
OK, but then how do we have getters in states.rs for those values?

indomitableSwan replied:
It would help if I used my own getters, of course.

indomitableSwan commented on zkabacus-crypto/src/revlock.rs:
```diff
@@ -16,55 +16,48 @@ use {
-#[derive(Debug, Serialize, Deserialize)]
-#[serde(try_from = "UncheckedRevocationSecret")]
-#[allow(missing_copy_implementations)]
+#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq)]
```
How do you make the functions `revocation_lock` and `revocation_secret` work if you don't have `Copy`?

yaymukund commented on zkabacus-crypto/src/revlock.rs:
```diff
@@ -87,42 +80,100 @@ pub struct RevocationLockCommitment(pub(crate) Commitment<G1Projective>);
+
+        // Determine if the result is a valid revocation lock,
+        // ie is a Scalar in canonical form (smaller than the modulus)
+        let maybe_lock = Scalar::from_bytes(&<[u8; 32]>::try_from(&digested[..]).unwrap());
```
The `GenericArray` type returned by `Digest::finalize()` encodes the length, so you can just do:

```rust
let maybe_lock = Scalar::from_bytes(digest.as_ref());
```

yaymukund commented on zkabacus-crypto/src/revlock.rs:
```diff
@@ -87,42 +80,100 @@ pub struct RevocationLockCommitment(pub(crate) Commitment<G1Projective>);
+        // ie is a Scalar in canonical form (smaller than the modulus)
+        let maybe_lock = Scalar::from_bytes(&<[u8; 32]>::try_from(&digested[..]).unwrap());
+
+        if maybe_lock.is_some().into() {
+            let valid_secret = RevocationSecret {
+                secret: unchecked.secret,
+                index: unchecked.index,
+            };
+            let valid_lock = RevocationLock(maybe_lock.unwrap());
+            Ok(RevocationPair {
+                secret: valid_secret,
+                lock: valid_lock,
+            })
+        } else {
+            Err(Error::InvalidSecret())
+        }
```
What do you think of:

```rust
match maybe_scalar.into() {
    Some(scalar) => Ok(RevocationPair {
        secret: RevocationSecret {
            secret: unchecked.secret,
            index: unchecked.index,
        },

        lock: RevocationLock(scalar),
    }),

    None => Err(Error::InvalidSecret),
}
```
?

Also, is it important that this particular conversion is done in constant time? I think when we use `into()` we lose this guarantee. See https://docs.rs/subtle/2.4.1/subtle/struct.CtOption.html#method.from

indomitableSwan replied:
I don't think it matters if this is constant time. @marsella?

marsella replied:
I think this is the point where the behavior changes depending on the value of the `CtOption`, so we have to break out of the constant time types.

indomitableSwan replied:
Concretely, what are you suggesting here?

marsella replied:
I think we should use Mukund's suggestion exactly as described above. This leaves the `into()`, which breaks the constant-time behavior of the `maybe_lock`. We don't need constant time guarantees here.

yaymukund commented on zkabacus-crypto/src/revlock.rs:
```diff
@@ -133,18 +184,13 @@ impl RevocationSecret {
 }
 
 impl RevocationLock {
-    /// Validate a revocation pair.
-    pub(crate) fn verify(&self, rs: &RevocationSecret) -> Verification {
-        Verification::from(self.0 == rs.revocation_lock().0)
-    }
-
     // Convert a revocation lock to its canonical [`Message`] representation.
-    pub(crate) fn to_message(&self) -> Message<1> {
+    pub(crate) fn to_message(self) -> Message<1> {
         Message::from(self.to_scalar())
     }
 
     /// Convert a revocation lock to its canonical `Scalar` representation.
-    pub(crate) fn to_scalar(&self) -> Scalar {
+    pub(crate) fn to_scalar(self) -> Scalar {
```
Good catch.

yaymukund commented on zkabacus-crypto/src/revlock.rs:
```diff
@@ -168,52 +214,78 @@ impl RevocationLockCommitment {
             parameters.revocation_commitment_parameters(),
             revocation_lock_blinding_factor.0,
-            &Message::from(revocation_lock.to_scalar()),
+            &Message::from(revocation_pair.lock.to_scalar()),
         );
 
-        match (pair_is_valid, opening_is_valid) {
-            (Verification::Verified, true) => Verification::Verified,
-            _ => Verification::Failed,
+        if opening_is_valid {
+            Verification::Verified
+        } else {
+            Verification::Failed
         }
```
Since we `impl From<bool> for Verification` you can do:

```rust
pub(crate) fn verify_revocation_pair(...) -> Verification {
    self.0.verify_opening(
      ...
    ).into()
}
```

yaymukund commented on zkabacus-crypto/src/revlock.rs:
```diff
@@ -222,20 +294,136 @@ mod test {
+        let mut rng = thread_rng();
+        for _ in 1..1000 {
+            let rp = RevocationPair::new(&mut rng);
+            let serialized_rp = bincode::serialize(&rp).unwrap();
+            let deserialized_rp = bincode::deserialize(&serialized_rp).unwrap();
```
This causes `cargo test` to error because `bincode` isn't available unless you enable the `sqlite` feature flag. To fix:

```rust
#[test]
#[cfg(feature = "sqlite")]
pub fn ...
```

yaymukund reviewed (changes requested):
I've left some comments inline, will finish reviewing tests tomorrow.

indomitableSwan commented on zkabacus-crypto/src/customer.rs:
```diff
@@ -346,6 +346,7 @@ pub struct Started {
 /// This is sent as part of zkAbacus.Pay.
 #[derive(Debug)]
 #[non_exhaustive]
+#[allow(missing_copy_implementations)]
```
I think this is a question for @marsella.

marsella replied:
I thought I was getting a warning when it wasn't there. Are you saying the default behavior is to not warn if `Copy`s are missing? 

It's missing because we don't want it to be easy to copy this and store it as an outdated channel status (which you could then use to try to close the channel).

indomitableSwan replied:
In any case, I think I'd like to add doc comments where we have a security reason for not implementing Copy. Otherwise I am not sure this decision would be lasting?

yaymukund replied:
> Are you saying the default behavior is to not warn if Copys are missing?

Yes, we override it https://github.com/boltlabs-inc/libzkchannels-crypto/blob/main/zkchannels-crypto/src/lib.rs#L9

indomitableSwan replied:
I suggest we make this an issue on its own and not part of this PR.

marsella replied:
That sounds like a good plan!

indomitableSwan commented on zkabacus-crypto/src/revlock.rs:
```diff
@@ -87,42 +80,100 @@ pub struct RevocationLockCommitment(pub(crate) Commitment<G1Projective>);
+
+    /// The provided revocation secret does not produce the input revocation lock.
+    #[error("The revocation secret does not produce the provided revocation lock")]
+    MismatchedPair(),
```
OK... after attempting to implement this, this looks like it would require increasing the visibility of the `UncheckedRevocationSecret`struct. After discussion with @marsella, returning the failing revocation secret here seems to have limited value...

yaymukund replied:
That makes sense. Thanks for looking into it. In that case, please remove the `()` for these dataless enum variants.

marsella reviewed (approved):
Looked through the tests more carefully, everything looks good to me. Thanks for all the iterating and for your patience in receiving reviews!

yaymukund reviewed (approved):
I left one non-blocking comment inline, but this lgtm

indomitableSwan commented:
Great! Thanks @yaymukund and @marsella for the helpful reviews. Should we wait to merge until we have the downstream PR for zeekoe ready?

indomitableSwan commented:
@marsella @yaymukund 
I made some additional changes to allow the customer to send a revocation pair to the merchant as part of the `LockMessage`, instead of a separate lock and secret.

I also updated `State`. Before the customer could retrieve the revocation secret on its own, and I changed it to retrieval of the full (lock, secret) pair.

marsella reviewed (approved):
LGTM!
